examples/nametoken/src/lib.rs

```diff
@@ -0,0 +1,1849 @@
+    fn from(le: TransferError) -> Self {
+        match le {
+            TransferError::AmountTooLarge => Self::AdminAmountTooLarge,
+            TransferError::MissingAccount => Self::MissingAdminAccount,
+        }
+    }
+}
+
+/// Mapping errors related to contract invocations to CustomContractError.
+impl<T> From<CallContractError<T>> for CustomContractError {
+    fn from(_cce: CallContractError<T>) -> Self { Self::InvokeContractError }
+}
+
+/// Mapping CustomContractError to ContractError
+impl From<CustomContractError> for ContractError {
+    fn from(c: CustomContractError) -> Self { Cis2Error::Custom(c) }
+}
+
+// Functions for creating, updating and querying the contract state.
+impl<S: HasStateApi> State<S> {
+    /// Creates a new state with no tokens.
+    fn empty(admin: AccountAddress, state_builder: &mut StateBuilder<S>) -> Self {
+        State {
+            admin,
+            state: state_builder.new_map(),
+            all_names: state_builder.new_map(),
+            implementors: state_builder.new_map(),
+        }
+    }
+
+    /// Register a name if it's not taken
+    fn register_fresh(
+        &mut self,
+        name: &ContractTokenId,
+        owner: &AccountAddress,
+        expires: Timestamp,
+        state_builder: &mut StateBuilder<S>,
+    ) -> ContractResult<()> {
+        let name_info = NameInfo::fresh(*owner, expires, state_builder);
+        // make sure that the name is not taken
+        ensure!(
+            self.all_names.insert(*name, name_info).is_none(),
+            CustomContractError::NameIsTaken.into()
+        );
+        let mut owner_state =
+            self.state.entry(*owner).or_insert_with(|| AddressState::empty(state_builder));
+        owner_state.owned_names.insert(*name);
+        Ok(())
+    }
+
+    /// Register a name if it's present in the registry, but expired
+    fn register_expired(
+        &mut self,
+        now: Timestamp,
+        name: &ContractTokenId,
+        owner: &AccountAddress,
+        state_builder: &mut StateBuilder<S>,
+    ) -> ContractResult<()> {
+        let name_info = self
+            .all_names
+            .get(name)
+            .ok_or(ContractError::Custom(CustomContractError::InconsistentState))?;
+        let old_expires = name_info.name_expires;
+        // check whether the name has expired
+        ensure!(now > old_expires, CustomContractError::NameIsTaken.into());
+        // transfer ownership
+        self.transfer(
+            name,
+            1.into(),
+            &Address::Account(name_info.owner),
+            &Address::Account(*owner),
+            state_builder,
+        )?;
+        let new_expires = now
+            .checked_add(Duration::from_days(REGISTRATION_PERIOD_DAYS))
+            .ok_or(ContractError::Custom(CustomContractError::InvokeContractError))?;
+        // update expiration date and replace old data with an empty vector
+        self.all_names
+            .get_mut(name)
```

Comment:
Ok, I've tried to just look up once and get a mutable reference, but now the compiler complains that `self.transfer` is trying to borrow `*self` as mutable, but it is borrowed as immutable when I look up for the first time
```rust
ln register_expired(
        &mut self,
        now: Timestamp,
        name: ContractTokenId,
        owner: AccountAddress,
        state_builder: &mut StateBuilder<S>,
    ) -> ContractResult<AccountAddress> {
        let mut name_info = self
            .all_names
            .get_mut(&name)
            .ok_or(ContractError::Custom(CustomContractError::InconsistentState))?;
        let old_expires = name_info.name_expires;
        // check whether the name has expired
        ensure!(now > old_expires, CustomContractError::NameIsTaken.into());

        let old_owner = name_info.owner;

        // transfer ownership
       // cannot borrow `*self` as mutable because it is also borrowed as immutable
        self.transfer(
            &name,
            1.into(),
            &Address::Account(old_owner),
            &Address::Account(owner),
            state_builder,
        )?;
        let new_expires = now
            .checked_add(Duration::from_days(REGISTRATION_PERIOD_DAYS))
            .ok_or(ContractError::Custom(CustomContractError::OverflowError))?;
        // update expiration date and replace old data with an empty vector
        name_info.name_expires = new_expires;
        name_info.data.replace(Vec::new());
        Ok(old_owner)
}
```

Reply:
No, I still run into the same problem.

Reply:
It would work, if the `self.transfer` call is the last statement. But in this case, I cannot return the old owner.

Reply:
If you add a `        drop(name_info)` before the self.transfer that should work.

Reply:
Also you should probably use `data.clear()` even though it is not technically the same.

Reply:
So, `data.clear()` removes all elements in the vector. But I use `StateBox`, would calling `clear()` require first loading the data? Is there a way of resetting the data to an empty vector without loading the previous content?

Reply:
`drop` fixed the issue, indeed.
For `data` I ended up with replacing it with a new vector and deleting old data. Potentially, it could be cheaper than loading the vector, but I haven't measured :)